### PR TITLE
Address lgtm.com warnings

### PR DESCRIPTION
**Pull request by zoidbergwill**
Closes https://github.com/git-multimail/git-multimail/issues/184

### CODE REVIEW

File: git-multimail/git_multimail.py [hunk 1]
```diff
@@ -777,6 +777,9 @@ def get_summary(self):
     def __eq__(self, other):
         return isinstance(other, GitObject) and self.sha1 == other.sha1
 
+    def __ne__(self, other):
+        return isinstance(other, GitObject) and self.sha1 != other.sha1
```
Comment (mhagger) [inline git-multimail/git_multimail.py]: This is not the inverse of `__eq__`, which would be

    return not isinstance(other, GitObject) or self.sha1 != other.sha1

or

    return not (self == other)

Note to self: This is actually only needed in Python2. In Python3, if there no `__ne__` is defined, then `__eq__` is called and its result is inverted.

File: git-multimail/git_multimail.py [hunk 2]
```diff
@@ -1503,10 +1503,9 @@ def generate_revision_change_summary(self, push):
                 new_commits_list = []
             new_commits = CommitSet(new_commits_list)
 
+            discarded_commits = CommitSet([])
```
Comment (mhagger) [inline git-multimail/git_multimail.py]: I agree that this is a false positive, even in Python3. I don't think we should worsen up the code to deal with a broken tool.

File: git-multimail/git_multimail.py [hunk 3]
```diff
@@ -2012,11 +2012,10 @@ def find_sendmail():
         for path in SendMailer.SENDMAIL_CANDIDATES:
             if os.access(path, os.X_OK):
                 return path
-        else:
-            raise ConfigurationException(
-                'No sendmail executable found.  '
-                'Try setting multimailhook.sendmailCommand.'
-                )
+        raise ConfigurationException(
```
Comment (mhagger) [inline git-multimail/git_multimail.py]: In both of these cases in this commit I prefer the original code because it alerts the reader that there is probably another `return` statement somewhere in the function.

**Comment by moy** [comment]
Hi,

Thanks for your interest in git-multimail. This sounds like a nice first contribution.

This lacks your sign-off (see https://github.com/git-multimail/git-multimail/blob/master/CONTRIBUTING.rst#contributing if not already done). This is why the Travis build fails.

File: git-multimail/git_multimail.py [hunk 4]
```diff
@@ -926,7 +926,7 @@ def generate_email_intro(self, html_escape_val=False):
 
         raise NotImplementedError()
 
-    def generate_email_body(self):
+    def generate_email_body(self, push):
```
Comment (moy) [inline git-multimail/git_multimail.py]: The change looks good, but as much as possible try to follow the following for the commit message:

```
1 short line (ideally <50 characters, and anyway <80) describing what is done, followed by a blank line

Text describing why the change is good, if needed.
```

File: git-multimail/git_multimail.py (hunk 2 above)
Comment (moy) [reply git-multimail/git_multimail.py]: I agree too. The lgtm people seem rather reactive so reporting the false positive may actually fix the issue (on the lgtm side).

File: git-multimail/git_multimail.py [hunk 5]
```diff
@@ -1978,7 +1978,7 @@ def __init__(self, environment, refname, short_refname, old, new, rev):
 class Mailer(object):
     """An object that can send emails."""
 
-    def __init__(self, environment):
+    def __init__(self, environment=None):
```
[the remark this answers is not in the document]
Comment (moy) [reply git-multimail/git_multimail.py]: Perhaps `OutputMailer` could accept an environment without requiring it (`environment=None`). That would avoid changing `Mailer` for a fix which is only about `OutputMailer`.

File: git-multimail/git_multimail.py [hunk 6]
```diff
@@ -3756,11 +3759,13 @@ def run_as_post_receive_hook(environment, mailer):
         changes.append(
             ReferenceChange.create(environment, oldrev, newrev, refname)
             )
-    if changes:
-        push = Push(environment, changes)
+    if not changes:
+        return
```
Comment (moy) [inline git-multimail/git_multimail.py]: When following this `return`, the method exits without closing the mailer. The connection is open in the constructor of `SMTPMailer`, not in the `send` method, so even if no mail is sent we still need to properly close the connection.

I don't think all this code change is needed. It should be sufficient to rename `__init__` to `close`, to re-add an `__init__` method that just calls `close`, and to change the call to `__init__` to `close`.

File: git-multimail/git_multimail.py [hunk 7]
```diff
@@ -3756,11 +3759,13 @@ def run_as_post_receive_hook(environment, mailer):
...
+    push = Push(environment, changes)
+    try:
         push.send_emails(mailer, body_filter=environment.filter_body)
-    if hasattr(mailer, '__del__'):
-        mailer.__del__()
+    finally:
```
Comment (moy) [inline git-multimail/git_multimail.py]: This try/finally is a good change, but ideally it would go to another commit. No strong objection if you prefer keeping it in the same commit though.

File: git-multimail/git_multimail.py [hunk 8]
```diff
@@ -2171,7 +2174,7 @@ def call(klass, server, timeout):
                 % (self.smtpserver, sys.exc_info()[1]))
             sys.exit(1)
 
-    def __del__(self):
+    def close(self):
```
Comment (moy) [inline git-multimail/git_multimail.py]: I think it's better to also keep a `__del__` method, so that any user of the class get the proper connection even without calling `close`. Not strictly needed, but it's just a 2-liners.

File: git-multimail/git_multimail.py (hunk 3 above)
Comment (moy) [reply git-multimail/git_multimail.py]: indeed, probably another false positive that could be reported to lgtm.

File: git-multimail/git_multimail.py (hunk 1 above)
Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: I was wondering about [`functools.total_ordering`](https://docs.python.org/3/library/functools.html#functools.total_ordering). I think the latter makes sense, and we can drop it one day if we ever drop Python 2 support, which this library probably won't.

File: git-multimail/git_multimail.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: Those both sound like good solutions, and cleaner than this one, and have less negative side effects.

File: git-multimail/git_multimail.py (hunk 3 above)
Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: Yeah, they have examples of this vs the earlier code for the rule, but I struggled to understand why one was objectively better than the other.

File: git-multimail/git_multimail.py (hunk 6 above)
Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: Yeah, some of this was just to match the other hook's structure more, but I can change it.

File: git-multimail/git_multimail.py (hunk 7 above)
Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: I can split it.

File: git-multimail/git_multimail.py (hunk 6 above)
Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: Hopefully the addition of call to close from  __del__ covers all those code paths now.

Comment (moy) [reply git-multimail/git_multimail.py]: > Hopefully the addition of call to close from `__del__` covers all those code paths now.

I don't think it does: `mailer` is passed as an argument to the function, so a `return` does not call the destructor.

Comment (zoidbergwill) [reply git-multimail/git_multimail.py]: Ahhh, that makes sense. I've never really used Python destructors before. Updating.
